src/adapter/src/coord.rs

```diff
@@ -1635,12 +1578,191 @@ impl Coordinator {
+                    let optimized_plan = global_mir_plan.df_desc().clone();
+
+                    // MIR ⇒ LIR lowering and LIR ⇒ LIR optimization (global)
+                    let global_lir_plan = optimizer.optimize(global_mir_plan)?;
+                    let physical_plan = global_lir_plan.df_desc().clone();
+                    let metainfo = global_lir_plan.df_meta().clone();
+
+                    let catalog = self.catalog_mut();
+                    catalog.set_optimized_plan(id, optimized_plan);
+                    catalog.set_physical_plan(id, physical_plan);
+                    catalog.set_dataflow_metainfo(id, metainfo);
+
+                    compute_instance.insert_collection(id);
+                }
+                _ => (),
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Collects for each index the materialized views that depend on it, either directly or
+    /// transitively through other indexes (but not through other MVs).
+    ///
+    /// The returned information is required during coordinator bootstrap for index as-of
+    /// selection, to ensure that selected as-ofs satisfy the requirements of downstream MVs.
+    ///
+    /// This method expects all dataflow plans to be available, so it must run after
+    /// `bootstrap_dataflow_plans`.
+    fn collect_index_dependent_matviews(&self) -> BTreeMap<GlobalId, BTreeSet<GlobalId>> {
+        // Collect imports of all indexes and MVs in the catalog.
+        let mut index_imports = BTreeMap::new();
+        let mut mv_imports = BTreeMap::new();
+        let catalog = self.catalog();
+        for entry in catalog.entries() {
+            let id = entry.id();
+            if let Some(plan) = catalog.try_get_physical_plan(&id) {
+                let imports: Vec<_> = plan.import_ids().collect();
+                if entry.is_index() {
+                    index_imports.insert(id, imports);
+                } else if entry.is_materialized_view() {
+                    mv_imports.insert(id, imports);
+                }
+            }
+        }
+
+        // Start with an empty set of dependants for each index.
+        let mut dependants: BTreeMap<_, _> = index_imports
+            .keys()
+            .map(|id| (*id, BTreeSet::new()))
+            .collect();
+
+        // Collect direct dependants first.
+        for (mv_id, mv_deps) in &mv_imports {
+            for dep_id in mv_deps {
+                if let Some(ids) = dependants.get_mut(dep_id) {
+                    ids.insert(*mv_id);
+                }
+            }
+        }
+
+        // Collect transitive dependants.
+        loop {
+            let mut changed = false;
+
+            // Objects with larger IDs tend to depend on objects with smaller IDs. We don't want to
+            // rely on that being true, but we can use it to be more efficient.
+            for (idx_id, idx_deps) in index_imports.iter().rev() {
```

Comment:
I wonder if there isn't an easier definition of this loop that only iterates over `dependants`. In theory this should improve the asymptotic bound of the worst case (linear dependency chain), but this is probably not so important at the moment.

Reply:
I think in practice today it will always be the case that dependencies have smaller IDs than their dependants*, so I'd expect the outer loop to be executed twice every time (the second time not making any changes and detecting the fixpoint). So I'm not really worried about the worst case behavior.

I don't think iterating over only `dependants` works, because that just contains `idx <- mv` dependency relations. But to find transitive dependencies of MVs on indexes (e.g. `idx1 <- idx2 <- mv`) we also need `idx1 <- idx2` relation information, so we need to look that up in `index_imports`. That's not to say that you couldn't iterate differently to end up with something more efficient, but I'm not sure if the code would be easier. Happy to be convinced of the opposite and find a way to simplify the logic here though!

*: Things get a bit weird with different `GlobalId` variants. It looks like they are ordered as `System < User < Transient`, which is also the only direction things can depend on one another. But it definitely feels brittle, so the code takes care to not rely on any of this.